### ISSUE TITLE
Tracking Issue for feature(unix_socket_ancillary_data)

### ISSUE THREAD

**Issue by LinkTed** [rust-lang/rust#76915]
Tracking issue for `feature(unix_socket_ancillary_data)` to extend  `UnixStream` and `UnixDatagram` to send and receive ancillary data. For example file descriptors.

### Unresolved Questions
Review the `SocketAncillary` struct before stabilization to ensure it is compatible with all OSes.

### Implementation history

PR #69864

**Comment by tadeokondrak** [comment]
`send_vectored_with_ancillary` should switch to non-mutable references for its parameters, since `sendmsg` doesn't write through them.

https://github.com/rust-lang/rust/pull/79753 changes just the `bufs` parameter, but it was closed for inactivity.

**Comment by LinkTed** [comment]
@tadeokondrak  I open a new PR #82589.

**Comment by LinkTed** [comment]
I create a new PR to send and receive TTL for UdpSocket from the ancillary interface. #82858

**Comment by reyk** [comment]
Hi @LinkTed, I posted a fix in https://github.com/rust-lang/rust/pull/83374 for macOS, OpenBSD, and possibly other BSDs.

**Comment by the8472** [comment]
I think the API still needs polish. All those niche socket features are tricky and right now it's probably not extensible enough for 3rd party libraries to fill platform-specific gaps such as flags or cmsgs only supported on one particular OS. These basically are wrappers for the `recvmsg` and `sndmsg` syscalls, we should expose them in a way that's forwards-compatible so we don't have to add `recv_vectored_with_ancillary_from_with_flags` or something like that in the future.

Before stabilizing we should consider whether `recv_vectored_with_ancillary_from` should gain another argument. I would recommend adding a `ReceiveOptions` argument that allows one to specify flags, similar to `OpenOptions`. The options could then be used to set linux' `MSG_DONTWAIT`, `MSG_PEEK` or `MSG_ERRQUEUE` for example, maybe via `custom_flags()` similar to `OpenOptions`. That way we don't want to support all OS-specific peculiarities.

Or we could put the flags into `SocketAncillary` or at least reserve the right to do that in the future, but in that case the struct would need to be renamed because it would serve several purposes at once.

Similar could be done for sending.

Additionally the `AncillaryData` enum should be marked as `#[non_exhaustive]` and maybe gain a `Unknown` variant for things that the standard library doesn't support. And a way to get the raw ancillary message. That way one could for example decode `struct sock_extended_err` in a 3rd party crate without waiting for std to implement it.

**Comment by LinkTed** [comment]
@the8472 Thank you for your feedback. If this PR request is #82858 merged I will start implementing your suggestions.

**Comment by tv42** [comment]
What about CMSG_SPACE? Without it, we're forced to either guess or probe the appropriate buffer size.

https://man7.org/linux/man-pages/man3/cmsg.3.html SCM_RIGHTS example shows how it's used to compute the correct buffer size (with some tricky business wrt alignment).

**Comment by LinkTed** [comment]
Hi @tv42, 
what are the difference between `CMSG_SPACE` and `CMSG_LEN`?

Is there a function, which calculate the alignment for a struct to a specific address in Rust?

**Comment by comex** [comment]
> what are the difference between `CMSG_SPACE` and `CMSG_LEN`?

CMSG_SPACE includes alignment bytes, CMSG_LEN doesn't.

> Is there a function, which calculate the alignment for a struct to a specific address in Rust?

I don't know what this means.

**Comment by LinkTed** [comment]
> > what are the difference between `CMSG_SPACE` and `CMSG_LEN`?
> 
> CMSG_SPACE includes alignment bytes, CMSG_LEN doesn't.
> 
> > Is there a function, which calculate the alignment for a struct to a specific address in Rust?
> 
> I don't know what this means.

I thought that the alignment depends on the address, where you want to store the struct.

**Comment by chloekek** [comment]
CMSG_ALIGN only takes a size argument, so the alignment only depends on the size, not on the address.

**Comment by chloekek** [comment]
What it means for CMGS_LEN to include “alignment bytes” is that it rounds the size up to include a number of padding bytes to make sure the next header is properly aligned (to the alignment of cmsghdr). This is needed because cmsghdr structs are tightly packed next to one another.

**Comment by LinkTed** [comment]
So, the address have to be a multiple of the return value of `CMSG_ALIGN`, right?

**Comment by chloekek** [comment]
The address of the cmsghdr object has to be a multiple of `std::men::align_of::<cmsghdr>()`.

**Comment by LinkTed** [comment]
Sorry, that I ask again. Does the macro `CMSG_ALIGN` the same as `std::mem::align_of::<cmsghdr>()`?

**Comment by chloekek** [comment]
`CMSG_ALIGN(len)` rounds len up to a multiple of `std::mem::align_of::<cmsghdr>()`.

This way when you have multiple cmsghdr structures side by side (with their payloads in between), each one is properly aligned.

**Comment by LinkTed** [comment]
Thanks, @chloekek. Currently, the code is using `CMSG_LEN`, which includes the alignment bytes. So, technically we have to ensure that the first `cmsg` struct is properly aligned, right? 

But, if `cmsg_len` is rounded up, this means that the receiver interpreted the alignment bytes as data. For example, if the sender set 3 bytes, which then is aligned to 8, the receiver will think that he got 8 bytes. How do we handle this?

**Comment by chloekek** [comment]
It seems that CMSG_NXTHDR itself adds the proper alignment:

https://github.com/spotify/linux/blob/6eb782fc88d11b9f40f3d1d714531f22c57b39f9/include/linux/socket.h#L116-L126

So cmsg_len does not have to be touched, just make sure to place the cmsghdr object at the correct address as indicated by CMSG_NXTHDR (or CMSG_ALIGN).

**Comment by chloekek** [comment]
Actually, the man page says that you _should_ include the padding in cmsg_len:

> CMSG_LEN() returns the value **to store in the cmsg_len member** of the cmsghdr structure

**Comment by LinkTed** [comment]
Interesting, currently there is a test, which send one FD and I tested on a 64-bit system, where the alignment of `cmsghdr` should be 8. 😀

**Comment by chloekek** [comment]
What is suspicious is that [CMSG_NXTHDR](https://github.com/torvalds/linux/blob/9d31d2338950293ec19d9b095fbaa9030899dcb4/include/linux/socket.h#L146) adds the alignment if necessary.

This could mean that _either_:

 - CMSG_NXTHDR is just being cautious to avoid undefined behavior (dereferencing pointer to unaligned data is UB).
 - You don’t actually have to include the padding in cmsg_len, despite what the man page suggests.

As for file descriptors, the kernel will indeed [divide the data length by the size of a file descriptor](https://github.com/torvalds/linux/blob/9d31d2338950293ec19d9b095fbaa9030899dcb4/net/core/scm.c#L72) in order to compute how many file descriptors to send. And it seems that you [indeed do have to set all of them to valid file descriptors](https://github.com/torvalds/linux/blob/9d31d2338950293ec19d9b095fbaa9030899dcb4/net/core/scm.c#L104-L105), not to `-1`.

**Comment by LinkTed** [comment]
I think the man page is wrong. [CMSG_LEN](https://github.com/torvalds/linux/blob/9d31d2338950293ec19d9b095fbaa9030899dcb4/include/linux/socket.h#L113) does not include the alignment. Only `CMSG_SPACE` as @comex suggested.

**Comment by chloekek** [comment]
The definition of CMSG_LEN in Linux is as follows:

```c
#define CMSG_LEN(len) (CMSG_ALIGN(sizeof(struct cmsghdr)) + (len))
```

It does not actually apply the rounding to the `len` argument, it applies the rounding to the size of the cmsghdr struct. Isn’t that redundant? sizeof is always a multiple of alignof.

**Comment by chloekek** [comment]
So to summarize:

 - Set cmsg_len to the return value of CMSG_LEN. cmsg_len can be any value greater than sizeof(cmsghdr).
 - Make sure that the next header is at a properly aligned address. CMSG_NXTHDR will round the address up for you, or you can use CMSG_ALIGN if you the pointer arithmetic yourself.

**Comment by LinkTed** [comment]
So, if `CMSG_NXTHDR` returns an aligned address, we have to ensure that the first `cmsghdr` is aligned because then a subsequent `cmsghdr` is aligned. Technically, we do not have to use `CMSG_SPACE`.

**Comment by LinkTed** [comment]
By the way @chloekek. Here is the commit, where the [`CMSG_ALIGN`](https://github.com/torvalds/linux/commit/1ff8cebf49ed9e9ca2ae44b5c4176aef9c21af9c) were removed.

**Comment by chloekek** [comment]
The SocketAncillary::new function currently takes any old `&[u8]`, regardless of alignment. Do we want to enforce its alignment? It could be made safe with an `assert!`ion on the address of the slice, but that would be rather annoying to use for users, as they would have to jump through hoops to make sure their buffer is properly aligned.

**Comment by LinkTed** [comment]
Yes, therefore I would add alignment into the `SocketAncillary::new ` function.

**Comment by chloekek** [comment]
That would mean that the actual length of the buffer can be 7 less than the length provided by the user (assuming an alignment of 8). Is this something we want? It is surprising behavior that would have to be documented well.

**Comment by LinkTed** [comment]
Unfortunately, this would not work if the user wants to use a dynamically allocated buffer, like `Vec<u8>`.

**Comment by LinkTed** [comment]
> Rounding up the address of the slice in SocketAncillary::new would mean that the actual length of the buffer can be up to 7 less ..

Unfortunately, yes 😞

**Comment by tv42** [comment]
> Rounding up the address of the slice in SocketAncillary::new would mean that the actual length of the buffer can be up to 7 less than the length provided by the user (assuming an alignment of 8). Is this something we want? It is surprising behavior that would have to be documented well.

Isn't this what `CMSG_SPACE` is for, to know how big of a buffer you need to allocate to fit the given amount of payload.

**Comment by tv42** [comment]
> So, if CMSG_NXTHDR returns an aligned address, we have to ensure that the first cmsghdr is aligned because then any subsequent cmsghdr is aligned. Technically, we do not have to use CMSG_SPACE.

The purpose of CMSG_SPACE is to know how large a buffer to allocate. Not to assist in filling it.

**Comment by tv42** [comment]
I think the road forward is going to look something like

```rust
        let num_fds = 7;
        let inner_size = (std::mem::size_of::<libc::c_int>() * num_fds) as libc::c_uint;
        let size = unsafe { libc::CMSG_SPACE(inner_size) } as usize;
        let align = std::mem::size_of::<libc::cmsghdr>() as usize;
        let align = align.next_power_of_two();
        let layout = std::alloc::Layout::from_size_align(size, align)
            .expect("alignment math for cmsghdr");
        let ptr = unsafe { std::alloc::alloc(layout) };
        let buf = unsafe { std::slice::from_raw_parts_mut(ptr, size) };
```

Wrapping that in a function in std::os::unix::net sounds like the right idea.

Maybe one for FDs, another for arbitrary data (taking just byte count, not FD count).

**Comment by tv42** [comment]
Err, lifetimes won't let you return a slice reference. Maybe returning `Box<[u8]>`? This is me running into limits of my knowledge of Rust idioms.

**Comment by tv42** [comment]
Something like this maybe:

```rust
pub struct SocketAncillary<'a> {
    buffer: &'a mut [u8],
    length: usize,
    truncated: bool,
}

impl<'a> SocketAncillary<'a> {
    pub fn new(buffer: &'a mut [u8]) -> Self {
        SocketAncillary {
            buffer,
            length: 0,
            truncated: false,
        }
    }

    pub fn with_capacity(size: usize) -> Self {
        let size = unsafe { libc::CMSG_SPACE(size as u32) } as usize;
        let align = std::mem::size_of::<libc::cmsghdr>() as usize;
        let align = align.next_power_of_two();
        let layout =
            std::alloc::Layout::from_size_align(size, align).expect("alignment math for cmsghdr");
        let ptr = unsafe { std::alloc::alloc(layout) };
        let buffer = unsafe { std::slice::from_raw_parts_mut(ptr, size) };
        Self::new(buffer)
    }

    pub fn with_capacity_for_fds(num: usize) -> Self {
        let size = std::mem::size_of::<libc::c_int>() * num;
        Self::with_capacity(size)
    }
}

fn main() {
    let s = SocketAncillary::with_capacity(42);
    println!("{:?}", s.buffer);
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=54898fddcec1ec9673c2658b92c3c8f1

**Comment by comex** [comment]
> ```
>        let align = std::mem::size_of::<libc::cmsghdr>() as usize;`
>        let align = align.next_power_of_two();
> ```
I wouldn't do this.  In practice this value should be enough, because on all of the implementations I've seen ([Linux](https://github.com/torvalds/linux/blob/9d31d2338950293ec19d9b095fbaa9030899dcb4/include/linux/socket.h#L92), [Darwin](https://github.com/apple/darwin-xnu/blob/main/bsd/sys/socket.h#L1107), [OpenBSD](https://github.com/openbsd/src/blob/master/sys/sys/socket.h), [Windows](https://github.com/tpn/winsdk-10/blob/master/Include/10.0.16299.0/shared/ws2def.h#L849)), the alignment that `CMSG_ALIGN` aligns to is <= 8, while the size of `cmsghdr` is >= 12.  But that's sort of a coincidence.  There's no real reason to think the size of `cmsghdr` has any relationship to the required alignment, especially given that platforms other than Linux still have macros such as `CMSG_SPACE` use `CMSG_ALIGN(sizeof(struct cmsghdr))` (or similar) rather than just `sizeof(struct cmsghdr)`.  (And on OpenBSD at least this actually makes a difference, since `sizeof(struct cmsghdr)` is always 12 while the required alignment is 8 on 64-bit targets.  In your proposed version, `next_power_of_two()` *would* cover that, ending up at a requested alignment of 16, but again, that's sort of a coincidence.)

On platforms which define `CMSG_ALIGN`, the required alignment should be equal to `CMSG_ALIGN(0)`.  Some platforms don't define it, but… since the `libc` bindings are maintained by hand anyway, I suggest just inspecting the definition of `CMSG_SPACE` on each platform to see what value it aligns to.

**Comment by chloekek** [comment]
All `CMSG_ALIGN` takes is the length. It won’t help with finding the address that the buffer should be positioned onto.

If you want to know the alignment for `cmsghdr`, can’t you just use `std::mem::align_of::<cmsghdr>()`? That should always give the correct answer. This is also what the union trick in the cmsg(3) man page implicitly uses (i.e. it uses the alignment as reported by the compiler, to position the buffer in memory).

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c2ef761bc1f42a99151c36c0ea535c9f

**Comment by chloekek** [comment]
If you want to heap-allocate the buffer, then this cannot be done by `SocketAncillary` itself, as `SocketAncillary` _borrows_ the buffer; it does not _own_ the buffer and so must not hold on to heap-allocated memory, as it would leak.

So if we want to offer an API for aligned buffers, it should be a separate structure:

```rust
pub struct SocketAncillaryBoxedBuf {
    // INVARIANT: Address is aligned to `align_of::<cmsghdr>()`.
    inner: Box<[u8]>,
}
impl SocketAncillaryBoxedBuf {
    pub fn with_capacity(size: usize) -> Self { /* aligned alloc here */ }
}
```

And replace `SocketAncillary::new` with:

```rust
impl<'a> SocketAncillary<'a> {
    pub fn from_boxed_buf(buffer: &mut SocketAncillaryBoxedBuf) -> Self { ... }
}

let mut example = SocketAncillaryBoxedBuf::with_capacity(12); // owns the memory
let mut ancillary = SocketAncillary::from_boxed_buf(&mut example); // borrows the memory
```

* * *

Then it would also be easy to add an interface for stack-allocated buffers:

```rust
pub union SocketAncillaryArrayBuf<const N: usize> {
    _align: cmsghdr,
    inner: [MaybeUninit<u8>; N],
}
impl<'a> SocketAncillary<'a>
    pub fn from_array_buf<const N: usize>(buffer: &mut SocketAncillaryArrayBuf<N>) -> Self { ... }
}

let mut example = SocketAncillaryArrayBuf<8>::new();
let mut ancillary = SocketAncillary::from_array_buf(&mut example);
```

**Comment by LinkTed** [comment]
And how the user can choose the bytes length of the buffer.

**Comment by comex** [comment]
> If you want to know the alignment for `cmsghdr`, can’t you just use `std::mem::align_of::<cmsghdr>()`? That should always give the correct answer. This is also what the union trick in the cmsg(3) man page implicitly relies on (i.e. it relies on rule that the alignment of a union type is the maximum alignment of its field types).

I am pretty sure the trick suggested in the Linux man page is actually wrong – at least morally speaking.  The alignment of `cmsghdr` is not necessarily sufficient.  For example, on OpenBSD (and FreeBSD, which I just verified does the same as OpenBSD here), `cmsghdr` is defined as

```c
struct cmsghdr {
    socklen_t   cmsg_len;
    int     cmsg_level;
    int     cmsg_type;
};
```

`socklen_t` is defined as `uint32_t`, so `struct cmsghdr` always has a size of 12 and alignment of 4.

However, the `CMSG_SPACE`, `CMSG_LEN`, and `CMSG_ALIGN` macros all use `_ALIGN`:

```c
#define CMSG_SPACE(l)       (_ALIGN(sizeof(struct cmsghdr)) + _ALIGN(l))
#define CMSG_LEN(l)     (_ALIGN(sizeof(struct cmsghdr)) + (l))
#define CMSG_ALIGN(n)   _ALIGN(n)
```

and at least on x86-64, `_ALIGN` rounds to 8 bytes:
```c
typedef	__int64_t	__register_t;
#define	_ALIGNBYTES	(sizeof(__register_t) - 1)
#define	_ALIGN(p)	(((__uintptr_t)(p) + _ALIGNBYTES) & ~_ALIGNBYTES)
```

Now, technically, the value that the *sizes* are aligned to doesn't have to be the same as the value the *pointers* are aligned to.  In practice, I believe (though haven't verified) that neither the BSDs nor Linux care about pointer alignment for cmsgs at all.  On all of those systems, `recvmsg`, `sendmsg`, etc. are implemented as system calls.  And on all of those systems, with cmsgs, like in most situations where userland passes pointers to the kernel, the kernel doesn't access the userland memory directly, but allocates its own buffer and effectively memcpys the data from/to userland as the first/last step.  And the kernel's allocated buffer has whatever alignment the kernel allocates it with, unrelated to the userland buffer's alignment.

But in theory those functions could be implemented in such a way that alignment matters, e.g. purely in userland, or in the kernel but with some mechanism where the kernel directly maps userland pages.  And while this is not explicitly stated, I would assume the pointer should be aligned to the same value the sizes are, because the whole point of aligning the sizes is to ensure that pointers stay aligned.

Unfortunately, there's no spec language to check this assumption against. The [POSIX man page for sys/socket.h](https://pubs.opengroup.org/onlinepubs/7908799/xns/syssocket.h.html) describes `cmsghdr` and *some* of the CMSG_* macros (namely, `CMSG_DATA`, `CMSG_NXTHDR`, and `CMSG_FIRSTHDR`), but it doesn't even mention `CMSG_SPACE` and `CMSG_LEN`, and it does not mention alignment at all.  Nor are those mentioned in the POSIX man pages for `recvmsg` and `sendmsg`.

**Comment by LinkTed** [comment]
Even if the function `recvmsg` and `sendmsg` is implemented in such a way that alignment does not matter, should be the `cmsghdr` be aligned, so that the user can access it? For example, to read the FDs.

To be honest, I do not see any big issue if the user allocates `BUFFER_SIZE + 7` bytes to ensure that there are enough bytes.

**Comment by tv42** [comment]
Another concern: Nothing seems to allow access to `recvmsg(2)` argument `flags`, or the similar return side `msghdr.msg_flags`.

Of these, specially interesting are

- `flags` `MSG_CMSG_CLOEXEC` prevents races with `fork(2)` when passing FDs
- `msghdr.msg_flags` `MSG_TRUNC` is very useful for correct error handling; otherwise you have to resort to kludges like using a buffer of size max_message_size+1 to tell the difference.

**Comment by LinkTed** [comment]
* `MSG_CMSG_CLOEXEC` is already set [see](https://github.com/rust-lang/rust/blob/e51830b90afd339332892a8f20db1957d43bf086/library/std/src/sys/unix/net.rs#L285)
* `MSG_TRUC` is also [returned](https://github.com/rust-lang/rust/blob/342db70ae4ecc3cd17e4fa6497f0a8d9534ccfeb/library/std/src/os/unix/net/ancillary.rs#L62), so the user can do error handling

**Comment by tv42** [comment]
Ah, I was reading https://doc.rust-lang.org/stable/std/os/unix/net/struct.UnixStream.html#method.recv_vectored_with_ancillary and https://doc.rust-lang.org/nightly/std/os/unix/net/struct.UnixStream.html#method.recv_vectored_with_ancillary are just `-> Result<usize>`. The `_from` variant is only on UnixDatagram. What I'm actually using is SOCK_SEQPACKET, because I do want a connection.

**Comment by LinkTed** [comment]
Is the `SOCK_SEQPACKET` supported by Rust's std, currently?

**Comment by the8472** [comment]
No, there was #50348 but it was recommended to implement it in an external crate first.

That said if https://github.com/rust-lang/rust/issues/76915#issuecomment-820682675 is addressed then I think one could make one of the unix sockets dual-purpose (only the method of construction would differ) since the EOR flag could be accessed through the extended API.

**Comment by tv42** [comment]
The semantics of SOCK_SEQPACKET depend on the domain. For UNIX domain sockets, they connect like SOCK_STREAM and send/receive like a (pseudo-connected) SOCK_DATAGRAM. If you FromRawFd an existing SOCK_SEQPACKET fd into a UnixDatagram, it works just fine.

**Comment by beroal** [comment]
> No, it is not buggy. You forgot to clear the buffer by `memset(cmsgs.data, 0, sizeof(cmsgs.data));`. If the next `cmsg` header has a bigger length than the `msg.control` then a null pointer is returned. see https://elixir.bootlin.com/linux/latest/source/include/linux/socket.h#L146

But the third `struct cmsghdr` in my code is within the bounds, and I wrote `__cmsg->cmsg_len` before calling `CMSG_NXTHDR` on `__cmsg`! So the macro you referenced shouldn't return `NULL` prematurely even without zeroing `cmsgs.data`. BTW, I can't find anything about this zeroing in the documentation.

BTW, I found the problem and wrote an answer. I believe, on my computer a different macro is used. It reads an uninitialized memory region which I'm unable to initialize because I don't have the pointer to it if I'm restricted to using `CMSG_NXTHDR`.

**Comment by beroal** [comment]
> No, It should not overwrite a control message over an existing one.

You're right, my bad. ☺

> > Should the buffer passed as the argument to the `new` method be aligned? If yes, how can a caller allocate such a buffer?
> 
> Currently, it should be aligned, but I would make PR to check if it is aligned. I do not know how the caller can allocate such a buffer, but it not a big deal because (IMHO) you would only lose max 7 bytes.

On the other hand, I believe, the alignment is necessary just for writing `struct cmsghdr`. IMHO, if you replace
```
        (*previous_cmsg).cmsg_level = cmsg_level;
        (*previous_cmsg).cmsg_type = cmsg_type;
        (*previous_cmsg).cmsg_len = libc::CMSG_LEN(source_len) as _;
```
with
```
        std::ptr::write_unaligned(&mut (*previous_cmsg).cmsg_level, cmsg_level);
        std::ptr::write_unaligned(&mut (*previous_cmsg).cmsg_type, cmsg_type);
        std::ptr::write_unaligned(&mut (*previous_cmsg).cmsg_len, libc::CMSG_LEN(source_len) as _);
```
the alignment requirement may be dropped altogether.

**Comment by LinkTed** [comment]
@beroal I am using arch Linux, too. I zeroed the buffer and it worked.

**Comment by beroal** [comment]
> @beroal I am using arch Linux, too. I zeroed the buffer and it worked.

It's because `cmsg_len` is now zero, so `CMSG_ALIGN` sees that the size of the control message is zero.

Adding
```
((struct cmsghdr *)((unsigned char *)cmsg + CMSG_SPACE(sizeof(cmsg_fd))))->cmsg_len = 0;
```
before
```
cmsg = CMSG_NXTHDR(&msg, cmsg);
```
will make it work too.

This may be a bug in Glibc ([the file in their repository](https://sourceware.org/git?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/bits/socket.h;h=a011a8c0959b9970f9784f431713a17b4ef7c682;hb=HEAD)).

**Comment by beroal** [comment]
> We can add such a function. Do you have any thought how the function should be named?

@LinkTed Would you like to consider the pull request https://github.com/rust-lang/rust/pull/88425 ?

**Comment by beroal** [comment]
[The `CMSG_NXTHDR` bug was discovered earlier.](https://stackoverflow.com/a/33893759/793392)

**Comment by the8472** [comment]
> BTW, I can't find anything about this zeroing in the documentation.

https://man7.org/linux/man-pages/man3/cmsg.3.html#DESCRIPTION


          When initializing a buffer that will contain a series of
          cmsghdr structures (e.g., to be sent with sendmsg(2)), that
          buffer should first be zero-initialized to ensure the correct
          operation of CMSG_NXTHDR().

**Comment by beroal** [comment]
I see a conversion from `usize` to `u32` and backward:
```
    let source_len = if let Some(source_len) = source.len().checked_mul(size_of::<T>()) {
        if let Ok(source_len) = u32::try_from(source_len) {
            source_len
        } else {
            return false;
        }
    } else {
        return false;
    };

    unsafe {
        let additional_space = libc::CMSG_SPACE(source_len) as usize;
```
I suppose it's because the type signatures of `libc::CMSG_LEN` and `libc::CMSG_SPACE` differ from [cmsg(3)](https://man7.org/linux/man-pages/man3/cmsg.3.html). Shouldn't we fix `libc` first?

**Comment by beroal** [comment]
@the8472 thanks.

**Comment by LinkTed** [comment]
> > We can add such a function. Do you have any thought how the function should be named?
> 
> @LinkTed Would you like to consider the pull request #88425 ?

@beroal Looks good.

**Comment by LinkTed** [comment]
> source_len

It because `size_of` returns `usize` and `CMSG_SPACE` excepted an `u32`.

**Comment by beroal** [comment]
> > source_len
> 
> It because `size_of` returns `usize` and `CMSG_SPACE` excepted an `u32`.

I mean that the signature of C's `CMSG_SPACE` is `size_t CMSG_SPACE(size_t length)`. AFAIK, C's `size_t` corresponds to Rust's `usize`. If the signature of Rust's `CMSG_SPACE` was `fn(length: usize) -> usize`, then it would be closer to C and integer conversions (blowing up code and threatening us with panics) would be unnecessary.

**Comment by beroal** [comment]
I've started composing an example for `ancillary_data_size` and have reviewed existing examples. I guess I insert a usage of `ancillary_data_size` into every example. Now a buffer `[0; 128]` for `SocketAncillary` is used in every example which, as I understand, is chosen arbitrarily.

I don't understand the purpose of `fds1` and `fds2` in the [example for `SocketAncillary::clear`](https://doc.rust-lang.org/stable/std/os/unix/net/struct.SocketAncillary.html#method.clear). `fds1` and `fds2` aren't used anywhere. Should I delete them?

**Comment by beroal** [comment]
In the [example for `SocketAncillary::clear`](https://doc.rust-lang.org/stable/std/os/unix/net/struct.SocketAncillary.html#method.clear), there is the following sequence of calls:
```
sock.recv_vectored_with_ancillary(bufs, &mut ancillary);
ancillary.clear();
sock.recv_vectored_with_ancillary(bufs, &mut ancillary);
```
What's the purpose of `clear` here? As I see from the implementation of `recv_vectored_with_ancillary`, it completely overwrites `ancillary` anyway.

Actually, what's the purpose of `clear` in general? We can just drop `SocketAncillary` and create a new `SocketAncillary`.

**Comment by beroal** [comment]
I tried to make `ancillary_data_size` `const`. However, this requires `libc::CMSG_SPACE` to be `const` which requires the `const-extern-fn` feature of `libc` to be on. When I request this feature in `Cargo.toml` of `std`, the repository `rust` doesn't compile because it uses a **beta** compiler. I'm in a dead end. I guess I need to wait until `const-extern-fn` is stabilized.

**Comment by LinkTed** [comment]
> I tried to make `ancillary_data_size` `const`. However, this requires `libc::CMSG_SPACE` to be `const` which requires the `const-extern-fn` feature of `libc` to be on. When I request this feature in `Cargo.toml` of `std`, the repository `rust` doesn't compile because it uses a **beta** compiler. I'm in a dead end. I guess I need to wait until `const-extern-fn` is stabilized.

Yes, I try it, too. We have to wait until `const-extern-fn` is stabilized.

**Comment by reyk** [comment]
@comex:
> Unfortunately, there's no spec language to check this assumption against. The [POSIX man page for sys/socket.h](https://pubs.opengroup.org/onlinepubs/7908799/xns/syssocket.h.html) describes `cmsghdr` and _some_ of the CMSG_* macros (namely, `CMSG_DATA`, `CMSG_NXTHDR`, and `CMSG_FIRSTHDR`), but it doesn't even mention `CMSG_SPACE` and `CMSG_LEN`, and it does not mention alignment at all. Nor are those mentioned in the POSIX man pages for `recvmsg` and `sendmsg`.

The API is from the IPv6 sockets API as found in RFC 2292 and the successor RFC 3542:

https://datatracker.ietf.org/doc/html/rfc2292#section-4
https://datatracker.ietf.org/doc/html/rfc3542#section-5

Everything is described in the RFC, e.g. CMSG_SPACE

```
   socklen_t CMSG_SPACE(socklen_t length);

   This macro is new with this API.  Given the length of an ancillary
   data object, CMSG_SPACE() returns an upper bound on the space
   required by the object and its cmsghdr structure, including any
   padding needed to satisfy alignment requirements.  This macro can be
   used, for example, when allocating space dynamically for the
   ancillary data.  This macro should not be used to initialize the
   cmsg_len member of a cmsghdr structure; instead use the CMSG_LEN()
   macro.
```

and CMSG_LEN

```
   socklen_t CMSG_LEN(socklen_t length);

   This macro is new with this API.  Given the length of an ancillary
   data object, CMSG_LEN() returns the value to store in the cmsg_len
   member of the cmsghdr structure, taking into account any padding
   needed to satisfy alignment requirements.

   Note the difference between CMSG_SPACE() and CMSG_LEN(), shown also
   in the figure in Section 20.2: the former accounts for any required
   padding at the end of the ancillary data object and the latter is the
   actual length to store in the cmsg_len member of the ancillary data
   object.
```

Linux' IPv6 stack was always different and not based on the KAME reference implementation which might explain the differences.

@LinkTed:
> Even if the function recvmsg and sendmsg is implemented in such a way that alignment does not matter, should be the cmsghdr be aligned, so that the user can access it? For example, to read the FDs.
> 
> To be honest, I do not see any big issue if the user allocates BUFFER_SIZE + 7 bytes to ensure that there are enough bytes.

OpenBSD is using file descriptor passing in almost all of its daemons and the common idiom to prepare a receive cmsg buffer for one optional file descriptor is the following:

```c
	union {
		struct cmsghdr hdr;
		char	buf[CMSG_SPACE(sizeof(int) * 1)];
	} cmsgbuf;
```

followed by:

```c       
	memset(&cmsgbuf, 0, sizeof(cmsgbuf));
 
	msg.msg_iovlen = 1;
	msg.msg_control = &cmsgbuf.buf;
	msg.msg_controllen = sizeof(cmsgbuf.buf);
```

**Comment by mdavidsaver** [comment]
As another data point, I am able to use `CMSG_SPACE()` in more complicated combinations of ipv4/6 control messages. Syntax is c++14.

```c++
// Linux/amd64 to receive SO_RXQ_OVFL and IP_PKTINFO or IPV6_PKTINFO
alignas (cmsghdr)
char cbuf[CMSG_SPACE(sizeof(int))+max(CMSG_SPACE(in_pktinfo), CMSG_SPACE(in6_pktinfo))];

// Some BSD derivatives to receive IP_ORIGDSTADDR + IP_RECVIF or IPV6_PKTINFO
alignas (cmsghdr)
char cbuf[max(CMSG_SPACE(sizeof(sockaddr_in))+CMSG_SPACE(sizeof(sockaddr_dl)), CMSG_SPACE(in6_pktinfo))];

// Windows to receive IP_PKTINFO or IPV6_PKTINFO
alignas (WSACMSGHDR)
char cbuf[max(WSA_CMSG_SPACE(in_pktinfo), WSA_CMSG_SPACE(in6_pktinfo))];
```

**Comment by m-ou-se** [comment]
Some of the standard library's tests do not compile on *BSD because of `AncillaryData` having a different definition that's missing the second variant on those platforms:

```
error: irrefutable `if let` pattern
   --> library/std/src/os/unix/net/tests.rs:657:8
    |
657 |     if let AncillaryData::ScmRights(scm_rights) = ancillary_data_vec.pop().unwrap().unwrap() {
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D irrefutable-let-patterns` implied by `-D warnings`
    = note: this pattern will always match, so the `if let` is useless
    = help: consider replacing the `if let` with a `let`

error: irrefutable `if let` pattern
   --> library/std/src/os/unix/net/tests.rs:778:8
    |
778 |     if let AncillaryData::ScmRights(scm_rights) = ancillary_data_vec.pop().unwrap().unwrap() {
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: this pattern will always match, so the `if let` is useless
    = help: consider replacing the `if let` with a `let`
```

**Comment by m-ou-se** [comment]
I'm starting to think that maybe we shouldn't have platform-specific APIs for tier 2 platforms. Those are not tested in CI. So in this case, the BSD platforms are the only ones where AncillaryData is an enum with a single variant. That leads to tons of errors in the tests, but is never tested in CI.

I'm opening a PR to disable this feature on BSD for now to get things back into a working state. If we want to bring this back on BSD we should make sure the API is not different on different platforms, as that's not really something `std` can or should do, and is very hard to maintain.

**Comment by m-ou-se** [comment]
This feature uses platform-specific `#[cfg]` attributes everywhere in its public API, which is something we currently cannot really do in the standard library. Right now we only expose such things through e.g. `os::unix` and through extension traits. Having extra methods appear on existing types on specific platforms, or having different definitions of the same enum on different platform is not something we can do in std today.

This API needs an RFC before we can consider working towards stabilizing it. Right now there is no overview of this API, not even in this tracking issue, and the current design does not fit all platforms, leading to issues like the one I mentioned in my previous comment.

**Comment by the8472** [comment]
I agree, the current state also has [other issues](https://github.com/rust-lang/rust/issues/76915#issuecomment-820682675)

Since the last PR (#82858) was closed due to inactivity it hasn't made any progress. It might be better to remove it and maybe develop it in `socket2` or as a separate crate instead and consider inclusion in std once it has been more polished.

**Comment by the8472** [comment]
Also, which BSD exactly is it failing on? If it's freebsd then it might be because there's an open PR for that? #91793

**Comment by joshtriplett** [comment]
I do care about the state of this API, and I'm in particular quite interested in the fd-sending mechanism.

It's certainly a good candidate for using whatever new `where Os: Unix` mechanism or similar that we introduce in the future.

**Comment by jmillikin** [comment]
I'm interested in this feature and have time to work on it. However, the current API is very different from what I would expect:
- The current API for `AncillaryData` is closely tied with UNIX sockets and doesn't seem to have any support for ancillary data from (for example) TCP/IPv6 sockets.
- Putting an enum for possible `SCM_*` values into `std::os::unix` prevents use of OS-specific values such as Linux's `SCM_SECURITY`.

How much flexibility is available for making changes to the API at this point? If I were to write an RFC with an API approximating the following, would that be a plausible next step toward stabilization?

```
// ===== API generic to any UNIX platform
trait unix::net::SocketSend {
  // equivalent to sendmsg()
  fn send_message(..., Option<&mut AncillaryData>) -> io::Result<usize>;
}
trait unix::net::SocketRecv {
  // equivalent to recvmsg()
  fn receive_message(..., Option<&mut AncillaryData>) -> io::Result<usize>;
}

struct unix::net::AncillaryData;
  fn control_messages() -> impl Iter<ControlMessage>;

struct unix::net::ControlMessage;
  fn level() -> c_int;
  fn type() -> c_int;
  fn data() -> &[u8]

// ===== Platform-specific API (Linux)
#[non_exhaustive]
enum linux::net::ControlMessage<'a> {
  ScmRights(ScmRights<'a>),
  ScmCredentials(ScmCredentials<'a>),
}
trait linux::net::ControlMessageExt: Sealed {
  fn try_get(&self) -> Option<linux::net::ControlMessage<'a>>,
}

struct linux::net::ScmCredentials<'a>;
  fn new(&unix::net::ControlMessage) -> Option<Self>;

struct linux::net::ScmRights<'a>;
  fn new(&unix::net::ControlMessage) -> Option<Self>;

// ===== Platform-specific API (etc)
trait freebsd::net::ControlMessageExt: Sealed { ... }
trait macos::net::ControlMessageExt: Sealed { ... }
trait netbsd::net::ControlMessageExt: Sealed { ... }
```

Advantages:
- API layout not affected by `#[cfg(target_os)]`
- New platform-specific functionality can be added freely via `non_exhaustive` enums.
- Exposing the `ControlMessage` fields directly allows users to self-service control message types not yet supported by the Rust standard library.

**Comment by kotauskas** [comment]
I found a potential critical concern with how `ScmRights` deals with its file descriptors upon being dropped – or, more specifically, how it does not.

[`rlimit` DoS via `SCM_RIGHTS`](https://gist.github.com/kentonv/bc7592af98c68ba2738f4436920868dc?permalink_comment_id=2876346#gistcomment-2876346) is a pretty insidious issue that can easily go unnoticed and needs to be manually mitigated by the user of the standard library to make sure that file descriptor table pollution is avoided and no resources are leaked.

My own implementation of Unix domain sockets and ancillary data in the [`interprocess`](https://github.com/kotauskas/interprocess) crate **(actual current ancillary data code not pushed to `main` yet)** does this differently and opts for a wrapper struct for `SCM_RIGHTS` (not too unlike `ScmRights`) that holds conditional ownership over its file descriptors, depending on whether it was constructed with &`[BorrowedFd]` or received from the kernel. If it's logically `&[OwnedFd]` with logical quasi-ownership through an immutable reference (as it would be if the FDs came from the kernel), it closes the descriptors upon being dropped, averting descriptor table pollution. It's not the cleanest implementation and I'm still working on it, but it's better than relegating prevention of `rlimit` DoS onto the downstream user and it takes care not to double-free.

**Comment by the8472** [comment]
Are you planning to implement it too? Asking since I intended to work on this eventually.

**Comment by jmillikin** [comment]
Yes, if that RFC is accepted then I plan to implement it. I already have it working in a local branch, so it'd just be a matter of adding docs and tests.

**Comment by jmillikin** [comment]
The RFC looks like it's moving now (tagged `libs-api-nominated`) so I've started work on the implementation of non-contentious parts.

* https://github.com/rust-lang/rust/pull/117156 is a change that needs to happen anyway, moving OS-specific socket methods to extension traits.
* https://github.com/rust-lang/rust/pull/117196 adds `ControlMessage{s,Iter,Buf}` for constructing and inspecting the control messages buffer used by ancillary data.

**Comment by purplesyringa** [comment]
As far as I can see, the current implementation is unsound: it takes a user-provided buffer and performs unaligned reads and writes. Is this a known issue?

**Comment by GrahamBarnett** [comment]
1. Would if be possible to make the constructor public?
  pub fn new(buffer: &'a mut [u8]), current_length: usize, truncated: bool -> Self {
    SocketAncillary { buffer, length: current_length, truncated }
  }
  this allows us to work on an existing ancillary buffer or re-use a SocketAncillary on different read buffers if we want, and as a by-product makes testing a lot easier
   
2. Please add #[derive(Debug)] to 
  SocketCred, 
  ScmRights, 
  AncillaryDataIter<'a, T>
  AncillaryData<'a>
  
  and:
  impl<'a> std::fmt::Debug for ScmRights<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let messages = Self {
            0: unsafe { AncillaryDataIter::<'a, RawFd>::new(self.0.data) },
        };
        f.debug_list().entries(messages.collect::<Vec<RawFd>>()).finish()
    }
  }
  
  impl<'a> std::fmt::Debug for ScmCredentials<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let messages = Self {
            0: unsafe { AncillaryDataIter::<'a, libc::ucred>::new(self.0.data) },
        };
        f.debug_list().entries(messages.collect::<Vec<SocketCred>>()).finish()
    }
  }
  
  impl<'a> std::fmt::Debug for Messages<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let messages = Messages{buffer: self.buffer, current: self.current};
        f.debug_list().entries(messages.collect::<Vec<Result<AncillaryData<'a>, AncillaryError>>>()).finish()
    }
  }
  
  this allows us to println!("{:?}", ancillary.messages()) which makes logging/ debugging a lot easier.
  It also means that you can debug the SocketAncillary struct itself (and get buffer status), or can show messages themselves as needed.
  
3. I had these changes suggested by the linter on 1.75:
    let data_len = (*cmsg).cmsg_len as usize - cmsg_len_zero; -->linter says should be--> let data_len = cmsg.cmsg_len - cmsg_len_zero;
    match (*cmsg).cmsg_level { -->linter says should be--> match cmsg.cmsg_level {
    cmsg_type: (*cmsg).cmsg_type, -->linter says should be--> cmsg_type: cmsg.cmsg_type,
  
Many thanks

**Comment by andrewbaxter** [comment]
> 1. Would if be possible to make the constructor public?

My use case is I don't know exactly where in a stream ancillary data will appear.  I'd like to use several reads to parse a message and aggregate ancillary data across them (in a single buffer) to be used at the end.  It doesn't appear trivially possible with the current design.

**Comment by GrahamBarnett** [comment]
I am doing exactly this. By making the constructor public and adding the
length and truncated to the constructor, I can customise the mio behaviour
as I need by managing the buffer/length/truncated collection myself using
the mio io call and only later creating this class to do the unpacking

**Comment by MaxVerevkin** [comment]
Wouldn't it be better for `SocketAncillary`'s `add_creds` and `add_fds` to return a `Result<(), NotEnoughSpaceErrror>` instead of a bool?

**Comment by jonleivent** [comment]
Would it be better for `SocketAncillary::new` to take a `MaybeUninit` buffer?  I assume that the original buffer should not be read from after calling `SocketAncillary::new`.  Also, this saves the initialization expense.

**Comment by the8472** [comment]
```
When initializing a buffer that will contain a series of
              cmsghdr structures (e.g., to be sent with [sendmsg(2)](https://man7.org/linux/man-pages/man2/sendmsg.2.html)),
              that buffer should first be zero-initialized to ensure the
              correct operation of CMSG_NXTHDR().
```

at least sending it requires zero-initialization anyway.

**Comment by jonleivent** [comment]
> that buffer should first be zero-initialized to ensure the
              correct operation of CMSG_NXTHDR().

Is that the responsibility of the caller of, or of `SocketAncillary::new` itself?  Because if this is a safety concern, how do you enforce it on the caller?  Wouldn't a common mistake be to re-use the buffer for subsequent `SocketAncillary::new` calls without resetting its contents?

**Comment by septatrix** [comment]
I see that it currently only includes `SO_PASSCRED`. What is the reason for that, is support for the other socket options also planned? It seems a bit arbitrary to only support one option but not the others (which are also frequently used)

**Comment by eesekaj** [comment]
Hello,
I would like to submit the following proposals for your consideration:

**The former:** 
In `SocketAncillary` a pointer or reference to inner buffer should be exposed and a buffer length should be modifiable r i.e 
`
fn get_ptr(&mut self) -> *mut u8;
// or
fn get_mut(&mut self) -> &mut [u8];
`
and 
`
fn set_len(&mut self, len: usize) -> Result<()>
`
Both changes will allow to use the `SocketAncillary` in the 3rd party code outside the STD, because at the moment I have to do bad things like:

```rust
let anci_struct: *mut SocketAncillaryCover = ancillary as *mut _ as *mut SocketAncillaryCover;

  if ancillary.len() > 0
  {
      msg.msg_controllen = ancillary.len();
      msg.msg_control = unsafe { (&mut *anci_struct).buffer.as_mut_ptr().cast() };
  }
```

[ancillary.rs](https://github.com/eesekaj/uds/blob/eae783d14dd7864dcc4a2f28ba82b8d02f434805/src/ancillary.rs#L759l)

**The second thing:**
In `SocketAncillary` in a function `add_fds` an argument fds accepts a slice of `RawFd` as input. In my humble (of cause) opinion, the argument should be &[OwnedFd], since once the FD(s) have been sent, it should no longer be accessible. 

At the moment, the code looks lile:
```rust
let fds = [OwnedFd::from(a).into_raw_fd(), OwnedFd::from(b).into_raw_fd(), OwnedFd::from(aa).into_raw_fd()];

ancillary.add_fds(&fds[..]);
```
But the `into_raw_fd()` function should be called inside the `add_to_ancillary_data()` consuming the OwnedFd.

I am continuing the maintenance of the crate "uds-fork" and those changes would suit the needs of the crate [ancillary.rs](https://github.com/eesekaj/uds/blob/master/src/ancillary.rs)

**Comment by Lancelotbronner** [comment]
What's the status of this feature, and what are the alternatives in the meantime?

**Comment by psychon** [comment]
@Lancelotbronner What exactly are you looking for? I think "all of this" is already possible using libc or rustix. For example, here is some code to [send](https://github.com/psychon/x11rb/blob/4cd9f2429a9e83d8963e569b977df79894f70cab/x11rb/src/rust_connection/stream.rs#L279-L322) and [receive](https://github.com/psychon/x11rb/blob/4cd9f2429a9e83d8963e569b977df79894f70cab/x11rb/src/rust_connection/stream.rs#L350-L388) FDs using rustix.

**Comment by Lancelotbronner** [comment]
Well for starters because the examples you've linked are a lot more code than what I got working with std, my only issue is that macOS support was removed.

Is there a source-compatible crate?

**Comment by Lancelotbronner** [comment]
What's the status on adding macOS support and standardizing this?
